### PR TITLE
[hapi] support dygraph amp O2

### PR DESCRIPTION
**Pull request by zhiqiu**
### PR types

New features
### PR changes

Others
### Describe

[hapi] support dygrapg amp O2

result of test_hapi_amp.py [since the loss can not get and check from model.fit()]

```
dynamic {'level': 'O1', 'init_loss_scaling': 128}
step 1/2 - loss: 314.2420
step 2/2 - loss: 286.8351
static {'level': 'O1', 'init_loss_scaling': 128}
step 1/2 - loss: 314.2420
step 2/2 - loss: 286.8125

dynamic {'level': 'O0'}
step 1/2 - loss: 314.2520 
step 2/2 - loss: 286.7574 
static {'level': 'O0'}
step 1/2 - loss: 314.2520 
step 2/2 - loss: 286.7574 

dynamic {'level': 'O2', 'init_loss_scaling': 128}
step 1/2 - loss: 314.2 
step 2/2 - loss: 285.8
static {'level': 'O2', 'init_loss_scaling': 128}
step 1/2 - loss: 314.2
step 2/2 - loss: 285.8
```

1. Model.fit() runs ok with Amp_level =  O0, O1 and O2 both in dynamic mode and static mode
2. The results in dynamic mode and static mode with Amp_level =  O0/O2 are equal, but slightly differs with Amp_level =  O1 since the O1 implementation in dynamic mode and static mode are not exactly equal.
3. The results with Amp_level =  O0, O1 and O2 are close (285.6282~286.8351), max diff is 0.4%

### CODE REVIEW

File: python/paddle/hapi/model.py [hunk 1]
```diff
@@ -455,10 +455,19 @@ def _run(self, inputs, labels=None):
...
         feed = {}
         input_names = [v.name for v in self._input_vars[self.mode]]
+        input_dtypes = [v.dtype for v in self._input_vars[self.mode]]
+
         for idx, n in enumerate(input_names):
             # train and test may take different arguments
             if inputs[idx] is not None:
                 feed[n] = inputs[idx]
+            if self._amp_level == 'O2' and input_dtypes[
+                    idx] == core.VarDesc.VarType.FP16:
```
Comment (zhangbo9674) [inline python/paddle/hapi/model.py:466]: is != core.VarDesc.VarType.FP16?

Comment (zhiqiu) [reply python/paddle/hapi/model.py:466]: input_dtypes is the expected dtype, we only perform cast when expected dtype is fp16.

File: python/paddle/hapi/model.py [hunk 2]
```diff
@@ -560,7 +569,11 @@ def _make_program(self, mode):
             inputs = [k._create_feed_layer() for k in to_list(inputs)]
             labels = [k._create_feed_layer() for k in to_list(labels)]
             self._label_vars[mode] = labels
-            outputs = to_list(self.model.network.forward(*inputs))
+            if self._amp_level == "O2" and core.is_compiled_with_cuda:
+                with paddle.static.amp.fp16_guard():
```
Comment (LiuChiachi) [inline python/paddle/hapi/model.py]: Is `use_fp16_guard` not needed in dygraph?

File: python/paddle/tests/test_hapi_amp.py [hunk 3]
```diff
@@ -26,34 +29,59 @@
...
+        optim = paddle.optimizer.Adam(
+            learning_rate=0.001, parameters=model.parameters())
+        model.prepare(
+            optimizer=optim,
+            loss=CrossEntropyLoss(reduction="sum"),
+            amp_configs=amp_config)
+        return model
+
+    def run_model(self, model):
+
```
Comment (LiuChiachi) [inline python/paddle/tests/test_hapi_amp.py]: Remove unnecessary blank lines:)

File: python/paddle/hapi/model.py [hunk 4]
```diff
@@ -1363,15 +1395,10 @@ def parameters(self, *args, **kwargs):
...
-            if self._adapter._amp_level == "O2":
-                if in_dygraph_mode():
-                    warnings.warn(
-                        "Pure float16 training is not supported in dygraph mode now, and it will be supported in future version."
-                    )
-                else:
-                    # grad clip is not supported in pure fp16 training now
-                    assert self._optimizer._grad_clip is None, \
-                        "Grad clip is not supported in pure float16 training now, and it will be supported in future version."
+            if self._adapter._amp_level == "O2" and self._optimizer._grad_clip:
```
Comment (LiuChiachi) [inline python/paddle/hapi/model.py:1408]: Maybe the doc of `Model` and `prepare()` also needs to be updated. Could you please update it:):):)

File: python/paddle/hapi/model.py [hunk 5]
```diff
@@ -804,17 +819,24 @@ def parameters(self, *args, **kwargs):
...
-
-    def load(self, param_state_pairs, optim_state):
+        if self.model._optimizer is not None:
+            if self.model._optimizer.state_dict():
+                optim = self.model._optimizer.state_dict()
+                fluid.save_dygraph(optim, path)
+        if hasattr(self.model, '_scaler') and self.model._scaler is not None:
+            if self.model._scaler.state_dict():
+                scaler = self.model._scaler.state_dict()
+                fluid.save_dygraph(scaler, path)
```
Comment (LiuChiachi) [inline python/paddle/hapi/model.py]: What's the suffix of scaler file? Maybe it's also`.pdparams`. Does scaler  need to be saved or loaded in static mode? I didn't add this before...

Comment (zhiqiu) [reply python/paddle/hapi/model.py]: It is needed as the optimizer state, plus I suggest upgrade the implementation and call `paddle.save/paddle.load` instead of `fluid.save_dygraph`.

File: python/paddle/hapi/model.py (hunk 2 above)
Comment (zhiqiu) [reply python/paddle/hapi/model.py]: yes, not needed

File: python/paddle/tests/test_hapi_amp.py (hunk 3 above)
Comment (zhiqiu) [reply python/paddle/tests/test_hapi_amp.py]: done

File: python/paddle/hapi/model.py (hunk 4 above)
Comment (zhiqiu) [reply python/paddle/hapi/model.py:1408]: done

**Review by LiuChiachi — APPROVED** [review]
LGTM

**Review by jzhang533 — APPROVED** [review]
lgtm
will update docs
